File: app/src/main/java/com/example/android/estados/PlotVanDerWaalsActivity.java
package com.example.android.estados;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GestureDetectorCompat;

import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.series.DataPoint;
import com.jjoe64.graphview.series.LineGraphSeries;

import java.util.ArrayList;

public class PlotVanDerWaalsActivity extends AppCompatActivity {

    private GestureDetectorCompat gestureDetectorCompat;
    private GraphView graph;

    private EditText ViEditText;
    private EditText VfEditText;
    private EditText PiEditText;
    private EditText PfEditText;
    private EditText TEditText;
    private EditText aEditText;
    private EditText bEditText;
    private EditText PartesEditText;

    int parts; // No. de partes de la gráfica
    int i, j; // Contadores
    final int iter = 100; //Máximo de iteraciones para el newton-raphson
    final double epsilon = 0.001; //Precisión de newton-raphson
    double a, b, T, P_i, P, P_f, delta, V, guess;
    final double cte_R = 0.082;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_plot_van_der_waals);

        // onClickListeners para los botones
//        TextView plotBtn = (TextView) findViewById(R.id.plot_btn);
//
//        plotBtn.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View view) {
//                graficar(view);
//            }
//        });

        TextView PlotButton = (TextView) findViewById(R.id.plot_btn);

        PlotButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                revisarCampos(view);
            }
        });

        TextView ResetButton = (TextView) findViewById(R.id.reset_btn);

        ResetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                borrarTodo(view);
            }
        });

        gestureDetectorCompat = new GestureDetectorCompat(this, new My2ndGestureListener());
        graph = (GraphView) findViewById(R.id.graph);

        ViEditText = (EditText) findViewById(R.id.Vi_edit_text);
        VfEditText = (EditText) findViewById(R.id.Vf_edit_text);
        PiEditText = (EditText) findViewById(R.id.Pi_edit_text);
        PfEditText = (EditText) findViewById(R.id.Pf_edit_text);
        TEditText = (EditText) findViewById(R.id.T_edit_text);
        aEditText = (EditText) findViewById(R.id.a_edit_text);
        bEditText = (EditText) findViewById(R.id.b_edit_text);
        PartesEditText = (EditText) findViewById(R.id.parts_edit_text);
    }

    // El siguiente código horroroso es para cambiar de pantalla
    @Override
    public boolean dispatchTouchEvent(MotionEvent ev){
        super.dispatchTouchEvent(ev);
        return gestureDetectorCompat.onTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        this.gestureDetectorCompat.onTouchEvent(event);
        return super.onTouchEvent(event);
    }

    class My2ndGestureListener extends GestureDetector.SimpleOnGestureListener {

        @Override
        public boolean onFling(MotionEvent event1, MotionEvent event2,
                               float velocityX, float velocityY) {

            if(event2.getX() > event1.getX()){
                Intent intent = new Intent(
                        PlotVanDerWaalsActivity.this, VanDerWaalsActivity.class);
                startActivity(intent);
                overridePendingTransition(R.anim.slide_right_out, R.anim.slide_right_in);
            }

            return true;
        }
    }

    ArrayList<Double> presion_data = new ArrayList<>(); // Valores de la presión
    ArrayList<Double> volum_data = new ArrayList<>(); // Valores del volumen
    LineGraphSeries<DataPoint> series = new LineGraphSeries<>(data());

    private void graficar(View view) {
        series.resetData(data());
        graph.addSeries(series);
    }

    public DataPoint[] data(){

        DataPoint[] values = new DataPoint[parts];
        for(i = 0; i < parts; i++){
            DataPoint v = new DataPoint(presion_data.get(i),volum_data.get(i));
            values[i] = v;
        }
        return values;
    }

    private void revisarCampos(View view) {

        // Revisar qué los campos están vacíos
        if (TextUtils.isEmpty(PartesEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            parts = Integer.valueOf(PartesEditText.getText().toString());
        }

        if (TextUtils.isEmpty(aEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            a = Double.valueOf(aEditText.getText().toString());
        }

        if (TextUtils.isEmpty(bEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            b = Double.valueOf(bEditText.getText().toString());
        }
        if (TextUtils.isEmpty(TEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            T = Double.valueOf(TEditText.getText().toString());
        }

        if (TextUtils.isEmpty(PiEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            P_i = Double.valueOf(PiEditText.getText().toString());
        }

        if (TextUtils.isEmpty(PfEditText.getText().toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show();
            return;
        } else {
            P_f = Double.valueOf(PfEditText.getText().toString());
        }

        delta = (P_f - P_i)/parts;
        llenarPresion(view);
        calcularVolumen(view);
        graficar(view);
    }

    private void llenarPresion(View view) {

        presion_data.clear();
        P = P_i;
        presion_data.add(P);

        for(i = 0; i < parts; i++) {
            P += delta;
            presion_data.add(P);
        }
    }

    private void calcularVolumen(View view) {

        volum_data.clear();
        P = P_i;

        for(i = 0; i < parts; i++) {
            guess = cte_R * T / P;

            //Aquí empieza el Newton-Raphson
            for (j = 0; j < iter; j++) {
                V = metodoVolumen(guess, P);
                if (Math.abs(V - guess) < epsilon) break;
                guess = V;
            }
            volum_data.add(V);
            P += delta;
        }
    }

    private double metodoVolumen(double x, double miP) {
        double f = (miP + (a / (x * x))) * (x - b) - (cte_R * T);
        double df = (miP + (a / (x * x))) + ((x - b) * (-2.0 * a / Math.pow(x, 3)));

        return x - (f / df);
    }

    private void borrarTodo(View view) {

        ViEditText.setText(null);
        VfEditText.setText(null);
        PiEditText.setText(null);
        PfEditText.setText(null);
        TEditText.setText(null);
        aEditText.setText(null);
        bEditText.setText(null);
        PartesEditText.setText(null);
    }

}

File: app/src/main/java/com/example/android/estados/PlotVanDerWaalsActivity.kt
package com.example.android.estados

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast

import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GestureDetectorCompat

import com.jjoe64.graphview.GraphView
import com.jjoe64.graphview.series.DataPoint
import com.jjoe64.graphview.series.LineGraphSeries

import java.util.ArrayList

class PlotVanDerWaalsActivity : AppCompatActivity() {

    private var gestureDetectorCompat: GestureDetectorCompat? = null
    private var graph: GraphView? = null

    private var ViEditText: EditText? = null
    private var VfEditText: EditText? = null
    private var PiEditText: EditText? = null
    private var PfEditText: EditText? = null
    private var TEditText: EditText? = null
    private var aEditText: EditText? = null
    private var bEditText: EditText? = null
    private var PartesEditText: EditText? = null

    internal var parts: Int = 0 // No. de partes de la gráfica
    internal var i: Int = 0
    internal var j: Int = 0 // Contadores
    internal val iter = 100 //Máximo de iteraciones para el newton-raphson
    internal val epsilon = 0.001 //Precisión de newton-raphson
    internal var a: Double = 0.toDouble()
    internal var b: Double = 0.toDouble()
    internal var T: Double = 0.toDouble()
    internal var P_i: Double = 0.toDouble()
    internal var P: Double = 0.toDouble()
    internal var P_f: Double = 0.toDouble()
    internal var delta: Double = 0.toDouble()
    internal var V: Double = 0.toDouble()
    internal var guess: Double = 0.toDouble()
    internal val cte_R = 0.082

    internal var presion_data = ArrayList<Double>() // Valores de la presión
    internal var volum_data = ArrayList<Double>() // Valores del volumen
    internal var series = LineGraphSeries(data())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_plot_van_der_waals)

        // onClickListeners para los botones
        //        TextView plotBtn = (TextView) findViewById(R.id.plot_btn);
        //
        //        plotBtn.setOnClickListener(new View.OnClickListener() {
        //            @Override
        //            public void onClick(View view) {
        //                graficar(view);
        //            }
        //        });

        val PlotButton = findViewById<View>(R.id.plot_btn) as TextView

        PlotButton.setOnClickListener { view -> revisarCampos(view) }

        val ResetButton = findViewById<View>(R.id.reset_btn) as TextView

        ResetButton.setOnClickListener { view -> borrarTodo(view) }

        gestureDetectorCompat = GestureDetectorCompat(this, My2ndGestureListener())
        graph = findViewById<View>(R.id.graph) as GraphView

        ViEditText = findViewById<View>(R.id.Vi_edit_text) as EditText
        VfEditText = findViewById<View>(R.id.Vf_edit_text) as EditText
        PiEditText = findViewById<View>(R.id.Pi_edit_text) as EditText
        PfEditText = findViewById<View>(R.id.Pf_edit_text) as EditText
        TEditText = findViewById<View>(R.id.T_edit_text) as EditText
        aEditText = findViewById<View>(R.id.a_edit_text) as EditText
        bEditText = findViewById<View>(R.id.b_edit_text) as EditText
        PartesEditText = findViewById<View>(R.id.parts_edit_text) as EditText
    }

    // El siguiente código horroroso es para cambiar de pantalla
    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        super.dispatchTouchEvent(ev)
        return gestureDetectorCompat!!.onTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        this.gestureDetectorCompat!!.onTouchEvent(event)
        return super.onTouchEvent(event)
    }

    internal inner class My2ndGestureListener : GestureDetector.SimpleOnGestureListener() {

        override fun onFling(event1: MotionEvent, event2: MotionEvent,
                             velocityX: Float, velocityY: Float): Boolean {

            if (event2.x > event1.x) {
                val intent = Intent(
                        this@PlotVanDerWaalsActivity, VanDerWaalsActivity::class.java)
                startActivity(intent)
                overridePendingTransition(R.anim.slide_right_out, R.anim.slide_right_in)
            }

            return true
        }
    }

    private fun graficar(view: View) {
        series.resetData(data())
        graph!!.addSeries(series)
    }

    fun data(): Array<DataPoint?> {

        val values = arrayOfNulls<DataPoint>(parts)
        i = 0
        while (i < parts) {
            val v = DataPoint(presion_data[i], volum_data[i])
            values[i] = v
            i++
        }
        return values
    }

    private fun revisarCampos(view: View) {

        // Revisar qué los campos están vacíos
        if (TextUtils.isEmpty(PartesEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            parts = Integer.valueOf(PartesEditText!!.text.toString())
        }

        if (TextUtils.isEmpty(aEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            a = java.lang.Double.valueOf(aEditText!!.text.toString())
        }

        if (TextUtils.isEmpty(bEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            b = java.lang.Double.valueOf(bEditText!!.text.toString())
        }
        if (TextUtils.isEmpty(TEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            T = java.lang.Double.valueOf(TEditText!!.text.toString())
        }

        if (TextUtils.isEmpty(PiEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            P_i = java.lang.Double.valueOf(PiEditText!!.text.toString())
        }

        if (TextUtils.isEmpty(PfEditText!!.text.toString())) {
            Toast.makeText(this, R.string.falta_algo, Toast.LENGTH_SHORT).show()
            return
        } else {
            P_f = java.lang.Double.valueOf(PfEditText!!.text.toString())
        }

        delta = (P_f - P_i) / parts
        llenarPresion(view)
        calcularVolumen(view)
        graficar(view)
    }

    private fun llenarPresion(view: View) {

        presion_data.clear()
        P = P_i
        presion_data.add(P)

        i = 0
        while (i < parts) {
            P += delta
            presion_data.add(P)
            i++
        }
    }

    private fun calcularVolumen(view: View) {

        volum_data.clear()
        P = P_i

        i = 0
        while (i < parts) {
            guess = cte_R * T / P

            //Aquí empieza el Newton-Raphson
            j = 0
            while (j < iter) {
                V = metodoVolumen(guess, P)
                if (Math.abs(V - guess) < epsilon) break
                guess = V
                j++
            }
            volum_data.add(V)
            P += delta
            i++
        }
    }

    private fun metodoVolumen(x: Double, miP: Double): Double {
        val f = (miP + a / (x * x)) * (x - b) - cte_R * T
        val df = miP + a / (x * x) + (x - b) * (-2.0 * a / Math.pow(x, 3.0))

        return x - f / df
    }

    private fun borrarTodo(view: View) {

        ViEditText!!.text = null
        VfEditText!!.text = null
        PiEditText!!.text = null
        PfEditText!!.text = null
        TEditText!!.text = null
        aEditText!!.text = null
        bEditText!!.text = null
        PartesEditText!!.text = null
    }

}
